library/sm3.c

```diff
@@ -0,0 +1,451 @@
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+/*
+ *  GB/T 32905-2016 compliant SM3 implementation
+ *  The SM3 algorithm was designed by Xiaoyun Wang et al in 2010.
+ *
+ *  http://www.gmbz.org.cn/upload/2018-07-24/1532401392982079739.pdf
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_SM3_C)
+
+#include "mbedtls/sm3.h"
+#include "mbedtls/platform_util.h"
+#include "mbedtls/error.h"
+
+#include <string.h>
+
+#if defined(MBEDTLS_SELF_TEST)
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#define mbedtls_printf printf
+#endif /* MBEDTLS_PLATFORM_C */
+#endif /* MBEDTLS_SELF_TEST */
+
+#if !defined(MBEDTLS_SM3_ALT)
+
+/*
+ * 32-bit integer manipulation macros (big endian)
+ */
+#ifndef GET_UINT32_BE
+#define GET_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (n) = ( (uint32_t) (b)[(i)    ] << 24 )             \
+        | ( (uint32_t) (b)[(i) + 1] << 16 )             \
+        | ( (uint32_t) (b)[(i) + 2] <<  8 )             \
+        | ( (uint32_t) (b)[(i) + 3]       );            \
+} while( 0 )
+#endif
+
+#ifndef PUT_UINT32_BE
+#define PUT_UINT32_BE(n,b,i)                            \
+do {                                                    \
+    (b)[(i)    ] = (unsigned char) ( (n) >> 24 );       \
+    (b)[(i) + 1] = (unsigned char) ( (n) >> 16 );       \
+    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
+    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
+} while( 0 )
+#endif
+
+void mbedtls_sm3_init( mbedtls_sm3_context *ctx )
+{
```

Comment:
```suggestion
{
       if( ctx == NULL )
        return;
```

Reply:
it cause Invalid parameter test fail .

Reply:
The “check_params” feature is going away in Mbed TLS 3.0. So please don't add any use of `MBEDTLS_INTERNAL_VALIDATE` and don't worry about a failure of a `check_params` test job.

Reply:
@gilles-peskine-arm @yuhaoth 
Can we agree that null pointers do not need to be treated any differently from any other invalid pointer as suggested by Gilles in https://github.com/ARMmbed/mbedtls/issues/4313, and therefore null pointer checks can  be removed from the code?
@yuhaoth Would you consider removing this specific parameter check from CI?

Reply:
Right, null pointers don't need to be treated differently. We are removing this check from the CI: it'll go away once https://github.com/ARMmbed/mbedtls/pull/4516 is merged.

Reply:
Per #4516 and gilles' comments , close this suggestion